tests/ignite/engine/test_engine.py

```diff
@@ -1279,6 +1279,44 @@ def check_iter_epoch(first_epoch_iter):
         assert engine.state.epoch == 10
         assert engine.state.iteration == 10 * real_epoch_length
 
+    def test_iterator_state_output(self):
+        torch.manual_seed(12)
+
+        def finite_iterator(length, batch_size):
+            for _ in range(length):
+                batch = torch.rand(batch_size, 3, 32, 32)
+                yield batch
+
+        def train_step(trainer, batch):
+            s = trainer.state
+            print(f"{s.epoch}/{s.max_epochs} : {s.iteration} - {batch.norm():.3f}")
+            return "flag_value"
+
+        trainer = Engine(train_step)
+        trainer.run(finite_iterator(4, 4), max_epochs=2)
+
+        assert trainer.state.output == "flag_value"
+        assert trainer.state.epoch == 2
+        # assert trainer.state.iteration == 2*4
```

Comment:
Found another bug: In the iterable case, the engine stops 1 epoch early, when its terminating due to StopIteration. 

Due to this comment, the test passes right now, but I haven't fixed the bug yet on this branch.

Reply:
I checked the test case and actually it is not a bug but kind of expected behavior with a finite-size input data iterator that was not manually restarted:
- `finite_iterator(4, 4)` can give only 4 batches and then StopIteration. This also defines the epoch length (as not provided in `.run(...)`). So, we increment the epoch counter 1 -> 2.
- As we do not restart the iterator manually, engine stops as it can't provide any data and also shows the following warning:
https://github.com/pytorch/ignite/blob/8d850e1f750281a14b304661eb5ff17dfbdad34f/ignite/engine/engine.py#L1114-L1118

Reply:
Thanks, I've played around with the test a bit, and I think I understand the explanation.  Ironically, had I fully understood that I'm supposed to reset the iterator in an event handler, I would have never written the code that found this bug in the first place!

I have a more readable version of the tests in a branch ( https://github.com/mtauraso/ignite/tree/issue/3372-test-followup ), with the correct checking on the commented-out line.

Should I make another PR since this one is merged?

Reply:
Now having read all comments I've put a PR here: https://github.com/pytorch/ignite/pull/3375